src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+-- loops.
+
+-- The Access represents accesses in "iteration space", i.e., the
+-- binders in the `ForEach` constructors correspond to loops in the
+-- user program.
+data Access (n::S) where
+  -- This chain of indices of this object is accessed this many times.
+  -- Corresponds to indexing.
+  Location :: [IxExpr n] -> Count -> Access n
+  -- The object may be accessed in any one of these ways (as, e.g., in
+  -- different arms of a downstream `case` expression).
+  -- Corresponds to case expressions.
+  Any :: [Access n] -> Access n
+  -- The object may be accessed in some or all of these ways (e.g.,
+  -- from multiple references).
+  -- Corresponds to sequential statements (e.g., x.i; x.j)
+  All :: [Access n] -> Access n
+  -- The object may be accessed in any of the ways given by the second
+  -- argument, as the name in the `Binder` argument varies over its
+  -- type.
+  -- Corresponds to `for`.
+  ForEach :: forall n l. (NameBinder AtomNameC n l) -> Access l -> Access n
+
+-- These are the indexing expressions we model.
+data IxExpr (n::S) =
+  -- Indexing by an identifier in scope
+    Var (Name AtomNameC n)
+  -- An index of product type
+  | Product [IxExpr n]
+  -- An index of sum type
+  | Inject Int (IxExpr n)
+  -- Unknown function of the given names
+  | Unknown [Name AtomNameC n]
+  -- All indices are accessed.  This constructor only appears after
+  -- `iterate`, below.
+  | IxAll
+  -- TODO Could add more constructors to capture other indexing
+  -- expressions, such as `from_ordinal _ $ ordinal i + 1`.  If that's
+  -- the only indexing expression the binding is safe to inline, and
+  -- inlining even preserves locality, not just work.
+  deriving (Eq, Show)
+
+-- We give Access semantics by a notional translation into a piece of
+-- code that dynamically counts the number of times every index is
+-- actually accessed.
+--
+-- - An Access becomes a function consuming a buffer reference
+--   and returning unit.  The buffer has the same type as the
+--   object whose accesses we are analyzing, except for integer
+--   leaves.
+--
+-- - For every control path of the source program, there is a control
+--   path of the Access program that accumulates at least as many
+--   counts into each element of the buffer as the number of times the
+--   source program read that element.
+--
+-- - `Location exprs ct` turns into accumulation
+--     \ref. ref!<exprs> += ct
+--     Note that this += must be understood as pointwise if
+--     the source indexing expression is not fully saturated.
+--
+-- - `Any opts` turns into case
+--     \ref. index = randint $ length opts
+--       case index of
+--         1 -> compile opt1 ref
+--         2 -> compile opt2 ref
+--         ...
+--
+-- - `All parts` turns into sequentialization
+--     \ref.
+--       compile part1 ref
+--       compile part2 ref
+--       ...
+--
+-- - `ForEach b sub` is iteration
+--     \ref. for b in (type of b).
```

Comment:
The `b` in `ForEach` is not annotated by its type, making this correspondence a little difficult to understand. Is it like that, because all `b`s are really instances of some index sets, but for the purpose of this analysis it doesn't matter which ones? If so, it might be worth adding that to the comment

Reply:
In the explanation I'm assuming that if necessary, I have the Dex environment lying around, where the type that `b` indexes can be looked up.  But the code currently doesn't use that type (only the semantics do), so there is no need to explicitly store it.  Updated the comments.